<a href="https://colab.research.google.com/github/jakekrafczyk/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [1]:
### YOUR CODE STARTS HERE

import numpy as np
import pandas as pd
from scipy import stats

names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names = names)
df.head()

--2020-02-18 00:11:50--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.4’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-02-18 00:11:50 (286 KB/s) - ‘house-votes-84.data.4’ saved [18171/18171]



party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [2]:
df = df.replace({"y": 1, "n": 0, "?": np.NaN})
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [3]:

columns = ['handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

#for col in columns:
 # df[col] = pd.to_numeric(df[col])


df.dtypes

party                    object
handicapped-infants     float64
water-project           float64
budget                  float64
physician-fee-freeze    float64
el-salvador-aid         float64
religious-groups        float64
anti-satellite-ban      float64
aid-to-contras          float64
mx-missile              float64
immigration             float64
synfuels                float64
education               float64
right-to-sue            float64
crime                   float64
duty-free               float64
south-africa            float64
dtype: object

In [4]:
rep = df[df['party']=='republican']
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [5]:
dem = df[df['party']=='democrat']
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [6]:
df.dtypes

party                    object
handicapped-infants     float64
water-project           float64
budget                  float64
physician-fee-freeze    float64
el-salvador-aid         float64
religious-groups        float64
anti-satellite-ban      float64
aid-to-contras          float64
mx-missile              float64
immigration             float64
synfuels                float64
education               float64
right-to-sue            float64
crime                   float64
duty-free               float64
south-africa            float64
dtype: object

In [7]:
#now run the two samp T test
stats.ttest_ind(rep['mx-missile'], dem['mx-missile'], nan_policy='omit')

Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-47)

In [8]:
print(rep['mx-missile'].mean())
dem['mx-missile'].mean()

0.11515151515151516


0.7580645161290323

Conclusion:

Based on a t-statistic of -16 and a p-value of ~0. I reject the null hypothesis that republicans and democrats support the 'mx-missile' bill at the same level. I suggest the alternative that democrat and republican levels of support for this bill are different, specifically that the democrats support the bill more than the republicans.

In [9]:
#now need to find a bill that the republicans support

stats.ttest_ind(rep['right-to-sue'],dem['right-to-sue'], nan_policy='omit')

Ttest_indResult(statistic=13.51064251060933, pvalue=1.2278581709672758e-34)

In [10]:
print(rep['right-to-sue'].mean())
dem['right-to-sue'].mean()

0.8607594936708861


0.2896825396825397

Conclusion:

Based on a t-statistic of 13 and a p-value of ~0. I reject the null hypothesis that republicans and democrats support the 'right-to-sue' bill at the same level. I suggest the alternative that democrat and republican levels of support for this bill are different, specifically that the republicans support the bill more than the democrats.

In [11]:
#now find an issue with only a small difference (p < 0.1)
stats.ttest_ind(rep['el-salvador-aid'], dem['el-salvador-aid'], nan_policy='omit')
#nope

Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68)

In [12]:
stats.ttest_ind(rep['aid-to-contras'], dem['aid-to-contras'], nan_policy='omit')
#nope

Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)

In [0]:

def ttest_all(columns):

  for col in columns:
    print(col,':',stats.ttest_ind(rep[col],dem[col],nan_policy='omit'),'\n')

In [14]:
ttest_all(columns)

handicapped-infants : Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18) 

water-project : Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485) 

budget : Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77) 

physician-fee-freeze : Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177) 

el-salvador-aid : Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68) 

religious-groups : Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20) 

anti-satellite-ban : Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31) 

aid-to-contras : Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54) 

mx-missile : Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-47) 

immigration : Ttest_indResult(statistic=1.7359117329695164, pvalue=0.08330248490425066) 

synfuels : Ttest_indResult(statistic=-8.29360398940

In [15]:
print(rep['water-project'].mean())
dem['water-project'].mean()

0.5067567567567568


0.502092050209205

Conclusion:

The water project bill is the only bill where we are not able to reject the null hypothesis that republicans and democrats support the bill at the same level. The T-value is 0.089 and the p-value is .929.